### PR TITLE
Add coverage measurement context

### LINKED ISSUE #195: Enable measurement context in coverage
Coverage rolled out measurement contexts not that long ago.  They are particularly useful when debugging coverage-related test failures because they let you see every test which exercised a line of code.

We'll need to add HTML reports to the coverage config, but I think that's a nice extra here.

### PR DESCRIPTION
**Pull request by ghickman**
This configures [Coverage's dynamic measurement contexts](https://coverage.readthedocs.io/en/6.4.4/contexts.html#dynamic-contexts) to help us find which tests executed a given line.

While adding that I also updated the black exclude config to ignore the output of coverage, and discovered the defaults are quite extensive (`/(\.direnv|\.eggs|\.git|\.hg|\.mypy_cache|\.nox|\.tox|\.venv|venv|\.svn|_build|buck-out|build|dist|__pypackages__)/` according to `black --help`!) allowing us to drop a couple of directories.

Fixes #195

### CODE REVIEW

**Review by meshy — APPROVED** [review]
You're on a roll!

File: pyproject.toml [hunk 1]
```diff
@@ -1,5 +1,13 @@
...
-extend-exclude = '/(\.direnv|\.env|\.venv|venv)/'
+extend-exclude = '''
+(
+  /(
+    | \.env
+    | \.pytest_cache
+    | htmlcov
+  )/
+)
+'''
```
[the remark this answers is not in the document]
Comment (ghickman) [reply pyproject.toml:10]: Oh wow, where has this setting been all my life?!  Happy to fix that up!

**Comment by ghickman** [comment]
@meshy – I've rebased the second commit with those file changes now being in the `.gitignore`.  I noticed bits of that file are a little stale but I'd like to puzzle over a few of those lines so do you mind if I do that in a separate change?

**Comment by meshy** [comment]
> I'd like to puzzle over a few of those lines so do you mind if I do that in a separate change?

Not at all! Thanks!

**Review by meshy — APPROVED** [review]

**Pull request merged**